File: 904-fruit-into-baskets/904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        // make unordered map of <int, int> representing type: amount
          std::unordered_map<int, int> m;

          // make int totalLen for storing global max length
          int totalLen = 0;
          int l = 0, r = 0;
          // while right is less than length of fruits array
          while(r < fruits.size()) {
              // add to unordered map and increment right
              m[fruits[r++]] += 1;
              // while number of elements in unordered map is greater than 2
              while(m.size() > 2) {
                // remove from left and increment left
                if(m[fruits[l]] == 1) {
                  m.erase(fruits[l]);
                } else {
                  m[fruits[l]]--;
                }
                l++;
              }
              // compare and take max length of previous and current
              totalLen = std::max(totalLen, r-l);
          }
          return totalLen;  
    }
};

File: 904-fruit-into-baskets/904-fruit-into-baskets.py
import collections

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        
        storage = collections.Counter()
        
        # make two pointers
        l, r, maxLen = 0, 0, 0
        
        while r < len(fruits):
            
            # add right pointer value into storage and increment right pointer
            storage[fruits[r]] += 1
            r += 1
            
            # move left pointer until distinct values in storage <= 2
            while len(storage) > 2: 
                if storage[fruits[l]] > 1: 
                    storage[fruits[l]] -= 1
                elif storage[fruits[l]] == 1: 
                    del storage[fruits[l]]
                l += 1
            
            maxLen = max(maxLen, r-l)
        
        return maxLen